fluent-dom/src/overlay.js

```diff
@@ -43,152 +43,148 @@ const LOCALIZABLE_ATTRIBUTES = {
+  // fromElement might be a {value, attributes} object as returned by
+  // Localization.messageFromContext. In which case attributes may be null to
+  // save GC cycles.
+  if (!fromElement.attributes) {
+    return;
+  }
 
-    for (const attr of safeAttributes) {
-      mergedChild.setAttribute(attr.name, attr.value);
+  // Set localizable attributes.
+  for (const attr of Array.from(fromElement.attributes)) {
+    if (isAttrNameLocalizable(attr.name, toElement, explicitlyAllowed)) {
+      toElement.setAttribute(attr.name, attr.value);
     }
-
-    translationFragment.replaceChild(mergedChild, childNode);
   }
-
-  // SourceElement might have been already modified by shiftNamedElement.
-  // Let's clear it to make sure other code doesn't rely on random leftovers.
-  sourceElement.textContent = "";
-
-  return translationFragment;
 }
 
 /**
- * Sanitize and merge attributes.
+ * Sanitize a child node created by the translation.
  *
- * Only localizable attributes from the translated child element and only
- * functional attributes from the source child element are considered safe.
+ * If childNode has the data-l10n-name attribute, try to find a corresponding
+ * child in sourceElement and use it as the base for the sanitization. This
+ * will preserve functional attribtues defined on the child element in the
+ * source HTML.
  *
- * @param {Element} translatedElement
- * @param {Element} sourceElement
- * @returns {Array<Attr>}
+ * @param   {Element} childNode - The child node to be sanitized.
+ * @param   {Element} sourceElement - The source for data-l10n-name lookups.
+ * @returns {Element}
  * @private
  */
-function sanitizeAttrsUsing(translatedElement, sourceElement) {
-  const localizedAttrs = Array.from(translatedElement.attributes).filter(
-    attr => isAttrNameLocalizable(attr.name, translatedElement)
-  );
+function sanitizeUsing(childNode, sourceElement) {
+  if (childNode.nodeType === childNode.TEXT_NODE) {
+    return childNode.cloneNode(false);
+  }
 
-  if (!sourceElement) {
-    return localizedAttrs;
+  if (childNode.hasAttribute("data-l10n-name")) {
+    const sourceChild = sourceElement.querySelector(
+      `[data-l10n-name="${childNode.getAttribute("data-l10n-name")}"]`);
+
+    if (sourceChild && sourceChild.localName === childNode.localName) {
+      // Remove it from sourceElement so that the translation cannot use
+      // the same reference name again.
+      sourceElement.removeChild(sourceChild);
+      // We can't currently guarantee that a translation won't remove
+      // sourceChild from the element completely, which could break the app if
+      // it relies on an event handler attached to the sourceChild. Let's make
+      // this limitation explicit for now by breaking the identitiy of the
+      // sourceChild by cloning it. This will destroy all event handlers
+      // attached to sourceChild via addEventListener and via on<name>
+      // properties.
+      const clone = sourceChild.cloneNode(false);
```

Comment:
I'd like to stay explicit in this iteration and try to remove this limitation in April. I think it should be possible to cache child elements in a WeakMap to prevent losing them to a bad translation. However, I'm not sure I fully understand the consequences of caching them like this and I also didn't want to make tests any more complex for now.